### PR TITLE
p2p event sync

### PR DESCRIPTION
**Pull request by CHr15F0x**
Fixes https://github.com/eqlabs/pathfinder/issues/1702

### CODE REVIEW

File: crates/p2p/src/client/peer_agnostic.rs [hunk 1]
```diff
@@ -500,6 +546,130 @@ impl Client {
         }
         }
     }
+
+    /// ### Important
+    ///
+    /// Events are grouped by block and by transaction. The order of flattened events in a block is guaranteed
+    /// to be correct because the event commitment is part of block hash. However the number of events per transaction
+    /// for __pre 0.13.2__ Starknet blocks is __TRUSTED__ because neither signature nor block hash contain this information.
```
Comment (CHr15F0x) [inline crates/p2p/src/client/peer_agnostic.rs:554]: I'm pretty sure this is correct... the [pre-0.13.2 signature](https://community.starknet.io/t/introducing-p2p-authentication-and-mismatch-resolution-in-v0-12-2/97993) does not cover it, neither does the block hash.

Comment (CHr15F0x) [reply crates/p2p/src/client/peer_agnostic.rs:554]: For example we just trust that for any pre-0.13.2 block:
```
(txn0, [eventA, eventB]),(txn1, [eventC])
```
is correct and not for example:
```
(txn0, [eventA]),(txn1, [eventB, eventC])
```

Comment (CHr15F0x) [reply crates/p2p/src/client/peer_agnostic.rs:554]: Ofc we won't have this ambiguity if the >=0.13.2 block hash could be computed and verified for pre 0.13.2 blocks.

File: crates/p2p/src/client/peer_agnostic.rs [hunk 2]
```diff
@@ -500,6 +546,130 @@ impl Client {
...
+                                        peer,
+                                        (start, std::mem::take(&mut events)),
+                                    );
+
+                                    // All the counters for this block have been exhausted which means
+                                    // that this block is complete.
+                                    if start < stop_inclusive {
+                                        // Move to the next block
+                                        start += 1;
+                                        current_count = event_counts_stream.next().await
```
[the remark this answers is not in the document]
Comment (CHr15F0x) [reply crates/p2p/src/client/peer_agnostic.rs:652]: `async_stream` macro makes the inner block immune from `fmt` :face_exhaling: , will fix this ofc

File: crates/storage/src/connection/block.rs [hunk 3]
```diff
@@ -509,6 +511,56 @@ pub(super) fn highest_block_with_all_class_definitions_downloaded(
...
+
+    while let Some(stat) = counts
+        .next()
+        .transpose()
+        .context("Iterating over event counts rows")?
+    {
+        ret.push(stat);
+    }
+
+    ret.reverse();
```
[the remark this answers is not in the document]
Comment (CHr15F0x) [reply crates/storage/src/connection/block.rs:559]: This is actually a bug because events are synced in block ascending order.

**Review by Mirko-von-Leipzig — APPROVED** [review]
LGTM modulo rebase

**Comment by CHr15F0x** [comment]
Test failures due to the feeder gateway 502s, unrelated to the PR's delta.

**Pull request merged**